### PR TITLE
fix(router): ensure routerLinkActive updates when associated routerLinks change

### LINKED ISSUE #18469: routerLinkActive not updating when routerLink changed
I have a routerLink URL that can change over time, but when the URL changes, routerLinkActive is not re-evaluated.

## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ x ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Change a routerLink URL to match/not match the current URL. routerLinkActive class is not applied/removed.

## Expected behavior
routerLinkActive is re-evaluated when its corresponding routerLink is updated.

## Minimal reproduction of the problem with instructions
Click "link a" or "link b" to make them active, then click the switchLinks button to change the URLs. Nothing changes. 
http://plnkr.co/edit/DHn4my9uAKJSIQSTIm1g?p=preview

## What is the motivation / use case for changing the behavior?
I have multiple departments managed by the same UI, with department name as a variable in the URL. When the user switches departments, and the on screen links or URL is changed, the UI doesn't update the routerLinkActive.

I think it's because ContentChildren.changes doesn't fire if the contents of a directive is changed, only when directives are added or removed.:
https://github.com/angular/angular/blob/master/packages/router/src/directives/router_link_active.ts#L106

Maybe it's too bad for performance to fix this?

## Environment

Angular version: 4.3
Mac

Browser:
- [x ] Chrome (desktop) version 59
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX

### PR DESCRIPTION
**Pull request by atscott**
This commit introduces a new subscription in the `routerLinkActive` directive which triggers an update
when any of its associated routerLinks have changes. `RouterLinkActive` not only needs to know when
links are added or removed, but it also needs to know about if a link it already knows about
changes in some way.

Quick note that `from...mergeAll` is used instead of just a simple
`merge` (or `scheduled...mergeAll`) to avoid introducing new rxjs
operators in order to keep bundle size down.

Fixes #18469

### CODE REVIEW

**Comment by atscott** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:325049401:BASE:325064353:1596652835642:658bbfd3)

File: packages/router/src/directives/router_link.ts [hunk 1]
```diff
@@ -175,6 +178,10 @@ export class RouterLink {
     }
   }
 
+  ngOnChanges(changes: SimpleChanges) {
+    this.onChanges.next();
```
Comment (AndrewKushnir) [inline packages/router/src/directives/router_link.ts:182]: I'd propose to add a comment here to mention why we emit here even though there is no additional logic in this lifecycle hook. I believe the reason is that the logic in corresponding getters is invoked, thus `ngOnChanges` was empty.

File: packages/router/src/directives/router_link_active.ts [hunk 2]
```diff
@@ -104,9 +104,22 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
 
 
   ngAfterContentInit(): void {
-    this.links.changes.subscribe(_ => this.update());
-    this.linksWithHrefs.changes.subscribe(_ => this.update());
-    this.update();
+    from([this.links.changes, this.linksWithHrefs.changes, of(null)])
```
Comment (AndrewKushnir) [inline packages/router/src/directives/router_link_active.ts:107]: Nice refactoring 👍 May be add a quick comment on why we have `of(null)` here?

File: packages/router/src/directives/router_link.ts [hunk 3]
```diff
@@ -175,6 +178,10 @@ export class RouterLink {
     }
   }
 
+  ngOnChanges(changes: SimpleChanges) {
```
Comment (AndrewKushnir) [inline packages/router/src/directives/router_link.ts:181]: FYI, this method will show up on https://angular.io/api/router/RouterLink, but this is similar to what [RouterLinkWithHref](https://angular.io/api/router/RouterLinkWithHref) has. I also noticed that on other Directives in other packages (like Forms), so I think we may need to come up with a strategy on how to represent these methods (and whether they are actually add any value in docs?) consistently for all directives, which is a topic for another PR(s) :)

**Review by AndrewKushnir — APPROVED** [review]
LGTM (just a couple nits) 👍 

Reviewed-for: public-api

File: packages/router/src/directives/router_link.ts (hunk 3 above)
Comment (atscott) [reply packages/router/src/directives/router_link.ts:181]: Maybe we can/should mark them with `/** @internal */`? As long as I'm making the changes in this PR, I might as well do that here if this seems like a good solution.

Comment (AndrewKushnir) [reply packages/router/src/directives/router_link.ts:181]: This sounds good to me, thanks Andrew 👍

**Review by mmalerba — APPROVED** [review]
LGTM

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

File: packages/router/src/directives/router_link_active.ts [hunk 4]
```diff
@@ -79,14 +80,13 @@ import {RouterLink, RouterLinkWithHref} from './router_link';
   exportAs: 'routerLinkActive',
 })
 export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit {
-  // TODO(issue/24571): remove '!'.
   @ContentChildren(RouterLink, {descendants: true}) links!: QueryList<RouterLink>;
-  // TODO(issue/24571): remove '!'.
```
Comment (petebacondarwin) [inline packages/router/src/directives/router_link_active.ts:84]: Generally I am against removing these comments unless the `!` is removed (somehow) or replacing the comment with a reason why the `!` can be retained.

**Review by alxhub — APPROVED** [review]
Reviewed-For: public-api

File: packages/router/src/directives/router_link_active.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/router/src/directives/router_link_active.ts:84]: But I guess Angular developers are aware that queries are only available after the appropriate hook has been called.

File: packages/router/src/directives/router_link.ts [hunk 5]
```diff
@@ -175,6 +178,13 @@ export class RouterLink {
     }
   }
 
+  /** @nodoc */
+  ngOnChanges(changes: SimpleChanges) {
+    // This is subscribed to by `RouterLinkActive` so that it knows to update when there are changes
+    // to the RouterLink's it's tracking.
```
Comment (petebacondarwin) [inline packages/router/src/directives/router_link.ts:184]:
```suggestion
    // to the RouterLinks it's tracking.
```

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: public-api

File: packages/router/src/directives/router_link_active.ts (hunk 4 above)
Comment (atscott) [reply packages/router/src/directives/router_link_active.ts:84]: My thought process was that they could "technically" be undefined until `afterContentInit` but are reliably initialized after that. And that we could really never make this `?` instead so the comment isn't really actionable in that way. I did some more reading in the issue report and found these guidelines that support keeping `!` as well:

> the field is reliably initialized, but TS cannot infer that (initialization can occur outside the ctor, for example), so ! should be kept.

https://github.com/angular/angular/issues/24571#issue-333470063
I think `ContentChildren` and `ViewChildren` could be considered "reliably initialized" after the correct lifecycle hooks.

> If the field is annotated with @ViewChildren or @ContentChildren - Add back '!' - c).

https://github.com/angular/angular/issues/24571#issuecomment-404606595

**Pull request closed without merging**